packages/@glimmer/runtime/test/new-initial-render-test.ts

```diff
@@ -0,0 +1,595 @@
+
+    this.rerender({ access: false });
+    this.assertHTML('<div><!---->!</div>');
+    this.assertStableNodes({ except: p });
+  }
+
+  @test "Loops"() {
+    this.render('<div>{{#each people key="handle" as |p|}}<span>{{p.handle}}</span> - {{p.name}}{{/each}}</div>', {
+      people: [
+        { handle: 'tomdale', name: 'Tom Dale' },
+        { handle: 'chancancode', name: 'Godfrey Chan' },
+        { handle: 'wycats', name: 'Yehuda Katz' }
+      ]
+    });
+
+    this.assertHTML('<div><span>tomdale</span> - Tom Dale<span>chancancode</span> - Godfrey Chan<span>wycats</span> - Yehuda Katz</div>');
+    this.assertStableRerender();
+
+    this.rerender({
+      people: [
+        { handle: 'tomdale', name: 'Thomas Dale' },
+        { handle: 'wycats', name: 'Yehuda Katz' }
+      ]
+    });
+
+    this.assertHTML('<div><span>tomdale</span> - Thomas Dale<span>wycats</span> - Yehuda Katz</div>');
+  }
+
+  protected compile(template: string): Template<Opaque> {
+    return this.env.compile(template);
+  }
+
+  render(template: string, properties: Dict<Opaque> = {}): void {
+    this.setProperties(properties);
+
+    this.renderResult = this.renderTemplate(this.compile(template));
+  }
+
+  protected abstract renderTemplate(template: Template<Opaque>): RenderResult;
+
+  rerender(properties: Dict<Opaque> = {}): void {
+    this.setProperties(properties);
+
+    this.env.begin();
+    expect(this.renderResult, 'the test should call render() before rerender()').rerender();
+    this.env.commit();
+  }
+
+  protected set(key: string, value: Opaque): void {
+    this.context[key] = value;
+  }
+
+  protected setProperties(properties: Dict<Opaque>): void {
+    Object.assign(this.context, properties);
+  }
+
+  protected takeSnapshot() {
+    let snapshot: (Node | 'up' | 'down')[] = this.snapshot = [];
+
+    let node = this.element.firstChild;
+    let upped = false;
+
+    while (node && node !== this.element) {
+      if (upped) {
+        if (node.nextSibling) {
+          node = node.nextSibling;
+          upped = false;
+        } else {
+          snapshot.push('up');
+          node = node.parentNode;
+        }
+      } else {
+        if (!isServerMarker(node)) snapshot.push(node);
+
+        if (node.firstChild) {
+          snapshot.push('down');
+          node = node.firstChild;
+        } else if (node.nextSibling) {
+          node = node.nextSibling;
+        } else {
+          snapshot.push('up');
+          node = node.parentNode;
+          upped = true;
+        }
+      }
+    }
+
+    return snapshot;
+  }
+
+  protected assertStableRerender() {
+    this.takeSnapshot();
+    this.runTask(() => this.rerender());
+    this.assertStableNodes();
+  }
+
+  protected assertHTML(html: string) {
+    equalTokens(this.element, html);
+  }
+
+  private runTask<T>(callback: () => T): T {
+    return callback();
+  }
+
+  protected assertStableNodes({ except: _except }: { except: Set<Node> | Node | Node[] } = { except: new Set() }) {
```

Comment:
We still support browsers without native `Set` (and we do not require polyfills).  We need to replace this with something else...